File: javachallange/src/main/java/com/example/darshanmistry/javachallange/LongestPrefixActivity.java
package com.example.darshanmistry.javachallange;

import android.os.Bundle;
import android.util.Log;
import androidx.appcompat.app.AppCompatActivity;

public class LongestPrefixActivity extends AppCompatActivity {

  @Override protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_main);

    //longestCommonPrefix(new String[] { "leets", "leetCode", "leet", "leeds" });
    String longPrefix =
        longestCommonPrefix(new String[] { "leets", "leetCode", "darshan", "demo" });
    Log.e("Java", "onCreate: " + longPrefix);
  }

  public String longestCommonPrefix(String[] strs) {
    if (strs.length == 0) return "";
    String prefix = strs[0];
    for (int i = 1; i < strs.length; i++)
      while (strs[i].indexOf(prefix) != 0) {
        prefix = prefix.substring(0, prefix.length() - 1);
        if (prefix.isEmpty()) return "";
      }
    return prefix;
  }
}

File: app/src/main/java/com/example/darshanmistry/kotlinchallange/LongestPrefixActivity.kt
package com.example.darshanmistry.kotlinchallange

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity

class LongestPrefixActivity : AppCompatActivity() {

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_main)

//    longestCommonPrefix(arrayOf("leets", "leetCode", "leeds", "leef"))
    binarySearchAlgo(arrayOf("leets", "leetCode", "leeds", "leef"))
  }

  private fun longestCommonPrefix(strs1: Array<String>): String {

    if (strs1.isEmpty()) {
      return ""
    }

    val sortedArray = strs1.sortedBy {
      it.length
    }
    var longPrefix = sortedArray[0]

    for (i in 1 until sortedArray.size) {
      val commonPrefix = sortedArray[i].commonPrefixWith(longPrefix, true)
      commonPrefix.length.takeIf { it > 0 }.apply {
        longPrefix = commonPrefix
      }
    }

    return longPrefix

  }

  private fun binarySearchAlgo(listStr: Array<String>) {
    val sortedArray = listStr.sortedBy {
      it.length
    }
    val binarySearch = sortedArray.binarySearch("leets", 0, sortedArray.size)
    println("---------Value---------------$binarySearch")

  }

}
